compiler/luci/pass/src/ResolveCustomOpMaxPoolWithArgmaxPass.cpp

```diff
@@ -0,0 +1,774 @@
+  auto const input = loco::must_cast<luci::CircleNode *>(cop->inputs(0));
+  CHECK_OR_FALSE(input->rank() == 4);
+
+  // TODO support batch size > 1 and `include_batch_in_index` option
+  CHECK_OR_FALSE(input->dim(0).value() == 1);
+
+  // Depth dimension is 3 because NHWC format is used
+  auto const depth_dimension = 3;
+
+  auto const input_depth = input->dim(depth_dimension).value();
+  auto const input_height = input->dim(1).value();
+  auto const input_width = input->dim(2).value();
+
+  // outputs
+  auto const outputs = loco::succs(cop);
+  assert(outputs.size() == cop->numOutputs());
+  CHECK_OR_FALSE(outputs.size() == 2);
+
+  auto output0 = loco::must_cast<luci::CircleCustomOut *>(*outputs.begin());
+  auto output1 = loco::must_cast<luci::CircleCustomOut *>(*(++outputs.begin()));
+
+  if (output0->index() > output1->index())
+    std::swap(output0, output1);
+
+  // From TF documentation: output of maxpool must has same type as input
+  assert(output0->dtype() == input->dtype());
+  assert(output1->dtype() == loco::DataType::S64);
+
+  assert(ksize_param.size() == input->rank());
+  assert(strides_param.size() == input->rank());
+
+  // Batch size and depth of ksize more than 1 is not supported.
+  CHECK_OR_FALSE(ksize_param[0] == 1 && ksize_param[3] == 1);
+  CHECK_OR_FALSE(strides_param[0] == 1 && strides_param[3] == 1);
+
+  // define Padding
+  auto padding = string_to_padding(padding_param);
+
+  // define Filter
+  luci::Filter filter;
+  filter.h(ksize_param[1]);
+  filter.w(ksize_param[2]);
+
+  // define Stride
+  luci::Stride stride;
+  stride.h(strides_param[1]);
+  stride.w(strides_param[2]);
+
+  // Create MaxPool
+  auto maxpool = graph->nodes()->create<luci::CircleMaxPool2D>();
+  {
+    maxpool->name(name + "/MaxPool2D");
+    luci::add_origin(maxpool, origin);
+
+    maxpool->stride()->h(stride.h());
+    maxpool->stride()->w(stride.w());
+
+    maxpool->filter()->h(filter.h());
+    maxpool->filter()->w(filter.w());
+
+    maxpool->padding(padding);
+
+    maxpool->fusedActivationFunction(luci::FusedActFunc::NONE);
+
+    // dtype of first output of MaxPoolWithArgmax is same as input's
+    maxpool->dtype(input->dtype());
+    define_max_pool_shape(input, maxpool);
+
+    maxpool->value(input);
+  }
+
+  // Create Split
+  auto split = graph->nodes()->create<luci::CircleSplit>();
+  {
+    split->name(name + "/Split");
+    luci::add_origin(split, origin);
+
+    // Create split_dim
+    auto split_dim = create_scalar<loco::DataType::S32>(graph, depth_dimension);
+    {
+      split_dim->name(split->name() + "/Dim");
+      luci::add_origin(split_dim, origin);
+    }
+
+    // Split input feature map to FM's with depth equal to 1
+    split->dtype(input->dtype());
+    copy_shape(input, split);
+    split->num_split(int32_t(input_depth));
+
+    split->split_dim(split_dim);
+    split->input(input);
+  }
+
+  // Create Concatenation
+  auto concat = graph->nodes()->create<luci::CircleConcatenation>(input_depth);
```

Comment:
Maybe move this to the bottom of the function?

Reply:
I created this node here in order to set `Concatenation` inputs (outputs of the branches) at the end of loop body:

https://github.com/Samsung/ONE/blob/ff94f1985fd76d0e4ca731d593644b3c2249c96d/compiler/luci/pass/src/ResolveCustomOpMaxPoolWithArgmaxPass.cpp#L722